File: C++/src/HashTable.cpp
#include "../include/HashTable.h"
#include <stdexcept>

HashTable::HashTable()
    : buckets(nullptr), capacity(16), size(0), loadFactor(0.75f) {
  buckets = new Node *[capacity];
  for (int i = 0; i < capacity; i++) {
    buckets[i] = nullptr;
  }
}

HashTable::HashTable(int initialCapacity)
    : buckets(nullptr), capacity(initialCapacity), size(0), loadFactor(0.75f) {
  if (capacity <= 0) {
    capacity = 16;
  }
  buckets = new Node *[capacity];
  for (int i = 0; i < capacity; i++) {
    buckets[i] = nullptr;
  }
}

HashTable::HashTable(const HashTable &other)
    : capacity(other.capacity), size(0), loadFactor(other.loadFactor) {
  buckets = new Node *[capacity];
  for (int i = 0; i < capacity; i++) {
    buckets[i] = nullptr;
  }

  for (int i = 0; i < other.capacity; i++) {
    Node *current = other.buckets[i];
    while (current != nullptr) {
      insert(current->key, current->value);
      current = current->next;
    }
  }
}

HashTable::~HashTable() {
  clear();
  delete[] buckets;
}

HashTable &HashTable::operator=(const HashTable &other) {
  if (this != &other) {
    clear();
    delete[] buckets;

    capacity = other.capacity;
    loadFactor = other.loadFactor;
    size = 0;

    buckets = new Node *[capacity];
    for (int i = 0; i < capacity; i++) {
      buckets[i] = nullptr;
    }

    for (int i = 0; i < other.capacity; i++) {
      Node *current = other.buckets[i];
      while (current != nullptr) {
        insert(current->key, current->value);
        current = current->next;
      }
    }
  }
  return *this;
}

int HashTable::hash(const std::string &key) const {
  unsigned long hash = 5381;
  for (char c : key) {
    hash = ((hash << 5) + hash) + c; // hash * 33 + c
  }
  return hash % capacity;
}

void HashTable::rehash() {
  int oldCapacity = capacity;
  Node **oldBuckets = buckets;

  capacity *= 2;
  buckets = new Node *[capacity];
  for (int i = 0; i < capacity; i++) {
    buckets[i] = nullptr;
  }

  int oldSize = size;
  size = 0;

  for (int i = 0; i < oldCapacity; i++) {
    Node *current = oldBuckets[i];
    while (current != nullptr) {
      insert(current->key, current->value);
      Node *temp = current;
      current = current->next;
      delete temp;
    }
  }

  delete[] oldBuckets;
}

void HashTable::insert(const std::string &key, const std::string &value) {
  if ((float)size / capacity >= loadFactor) {
    rehash();
  }

  int index = hash(key);
  Node *current = buckets[index];

  // Update if key exists
  while (current != nullptr) {
    if (current->key == key) {
      current->value = value;
      return;
    }
    current = current->next;
  }

  // Insert new node
  Node *newNode = new Node(key, value);
  newNode->next = buckets[index];
  buckets[index] = newNode;
  size++;
}

std::string HashTable::get(const std::string &key) const {
  int index = hash(key);
  Node *current = buckets[index];

  while (current != nullptr) {
    if (current->key == key) {
      return current->value;
    }
    current = current->next;
  }

  throw std::runtime_error("Key not found");
}

bool HashTable::contains(const std::string &key) const {
  int index = hash(key);
  Node *current = buckets[index];

  while (current != nullptr) {
    if (current->key == key) {
      return true;
    }
    current = current->next;
  }
  return false;
}

void HashTable::remove(const std::string &key) {
  int index = hash(key);
  Node *current = buckets[index];
  Node *prev = nullptr;

  while (current != nullptr) {
    if (current->key == key) {
      if (prev == nullptr) {
        buckets[index] = current->next;
      } else {
        prev->next = current->next;
      }
      delete current;
      size--;
      return;
    }
    prev = current;
    current = current->next;
  }
}

void HashTable::clear() {
  for (int i = 0; i < capacity; i++) {
    Node *current = buckets[i];
    while (current != nullptr) {
      Node *temp = current;
      current = current->next;
      delete temp;
    }
    buckets[i] = nullptr;
  }
  size = 0;
}

void HashTable::serialize(std::ofstream &out) const {
  out.write(reinterpret_cast<const char *>(&size), sizeof(size));

  for (int i = 0; i < capacity; i++) {
    Node *current = buckets[i];
    while (current != nullptr) {
      int keyLen = current->key.length();
      int valLen = current->value.length();

      out.write(reinterpret_cast<const char *>(&keyLen), sizeof(keyLen));
      out.write(current->key.c_str(), keyLen);

      out.write(reinterpret_cast<const char *>(&valLen), sizeof(valLen));
      out.write(current->value.c_str(), valLen);

      current = current->next;
    }
  }
}

void HashTable::deserialize(std::ifstream &in) {
  clear();
  int newSize;
  in.read(reinterpret_cast<char *>(&newSize), sizeof(newSize));

  for (int i = 0; i < newSize; i++) {
    int keyLen, valLen;

    in.read(reinterpret_cast<char *>(&keyLen), sizeof(keyLen));
    char *keyBuffer = new char[keyLen + 1];
    in.read(keyBuffer, keyLen);
    keyBuffer[keyLen] = '\0';

    in.read(reinterpret_cast<char *>(&valLen), sizeof(valLen));
    char *valBuffer = new char[valLen + 1];
    in.read(valBuffer, valLen);
    valBuffer[valLen] = '\0';

    insert(std::string(keyBuffer), std::string(valBuffer));

    delete[] keyBuffer;
    delete[] valBuffer;
  }
}

File: go/structs/hashtable/hashtable.go
package hashtable

import (
	"encoding/json"
	"fmt"
)

// Node represents a hash table entry
type Node struct {
	Key   string
	Value string
	Next  *Node
}

// HashTable represents a hash table with chaining
type HashTable struct {
	buckets    []*Node
	capacity   int
	size       int
	loadFactor float64
}

// New creates a new HashTable
func New() *HashTable {
	capacity := 16
	return &HashTable{
		buckets:    make([]*Node, capacity),
		capacity:   capacity,
		size:       0,
		loadFactor: 0.75,
	}
}

// hash computes hash value for a key
func (ht *HashTable) hash(key string) int {
	hash := 5381
	for _, c := range key {
		hash = ((hash << 5) + hash) + int(c)
	}
	if hash < 0 {
		hash = -hash
	}
	return hash % ht.capacity
}

// Insert adds or updates a key-value pair
func (ht *HashTable) Insert(key, value string) {
	if float64(ht.size)/float64(ht.capacity) >= ht.loadFactor {
		ht.rehash()
	}

	index := ht.hash(key)
	current := ht.buckets[index]

	// Update if key exists
	for current != nil {
		if current.Key == key {
			current.Value = value
			return
		}
		current = current.Next
	}

	// Insert new node
	newNode := &Node{Key: key, Value: value, Next: ht.buckets[index]}
	ht.buckets[index] = newNode
	ht.size++
}

// Get retrieves value for a key
func (ht *HashTable) Get(key string) (string, error) {
	index := ht.hash(key)
	current := ht.buckets[index]

	for current != nil {
		if current.Key == key {
			return current.Value, nil
		}
		current = current.Next
	}

	return "", fmt.Errorf("key not found: %s", key)
}

// Contains checks if key exists
func (ht *HashTable) Contains(key string) bool {
	_, err := ht.Get(key)
	return err == nil
}

// Remove deletes a key-value pair
func (ht *HashTable) Remove(key string) {
	index := ht.hash(key)
	current := ht.buckets[index]
	var prev *Node

	for current != nil {
		if current.Key == key {
			if prev == nil {
				ht.buckets[index] = current.Next
			} else {
				prev.Next = current.Next
			}
			ht.size--
			return
		}
		prev = current
		current = current.Next
	}
}

// rehash doubles capacity and rehashes all entries
func (ht *HashTable) rehash() {
	oldBuckets := ht.buckets
	ht.capacity *= 2
	ht.buckets = make([]*Node, ht.capacity)
	ht.size = 0

	for _, bucket := range oldBuckets {
		current := bucket
		for current != nil {
			ht.Insert(current.Key, current.Value)
			current = current.Next
		}
	}
}

// Size returns number of entries
func (ht *HashTable) Size() int {
	return ht.size
}

// IsEmpty checks if table is empty
func (ht *HashTable) IsEmpty() bool {
	return ht.size == 0
}

// Clear removes all entries
func (ht *HashTable) Clear() {
	ht.buckets = make([]*Node, ht.capacity)
	ht.size = 0
}

// ToMap returns all entries as map
func (ht *HashTable) ToMap() map[string]string {
	result := make(map[string]string)
	for _, bucket := range ht.buckets {
		current := bucket
		for current != nil {
			result[current.Key] = current.Value
			current = current.Next
		}
	}
	return result
}

// MarshalJSON implements json.Marshaler
func (ht *HashTable) MarshalJSON() ([]byte, error) {
	return json.Marshal(ht.ToMap())
}

// UnmarshalJSON implements json.Unmarshaler
func (ht *HashTable) UnmarshalJSON(data []byte) error {
	var m map[string]string
	if err := json.Unmarshal(data, &m); err != nil {
		return err
	}

	ht.Clear()
	for k, v := range m {
		ht.Insert(k, v)
	}
	return nil
}
